R/devmode.R

```diff
@@ -0,0 +1,251 @@
+#'
+#' If your package should not take a dependency on Shiny, we recommending re-implementing two functions:
+#'
+#' 1. `in_devmode()`: This function should return `TRUE` if `getOption("shiny.devmode")` is set. In addition, we strongly recommend that it also checks to make sure `testthat` is not testing.  This can be achieved using:
+#' ```r
+#' in_devmode <- function() {
+#'   isTRUE(getOption("shiny.devmode", FALSE)) &&
+#'     # !testthat::is_testing()
+#'     !identical(Sys.getenv("TESTTHAT"), "true")
+#' }
+#' ```
+#'
+#' 2. `get_devmode_option(option, default, on, message)`:
+#' This function should return the global `option` when it is set, similar to `getOption(option, default)`. When the global option `option` is not set, this function should return the default value (`default`) when `in_devmode()` is `FALSE` and the default Dev Mode value (`on`) when `in_devmode()` is `TRUE`. We strongly recommend displaying a message (`message`) to the developer once every 8 hours using `rlang::inform(message, .frequency = "regularly", .frequency_id = message)` if returning the `on` default Dev Mode option value. This will keep the author up to date as to which behaviors are being altered. To allow developers a chance to disable Dev Mode messages, the message may be skipped if `getOption("shiny.devmode.verbose", TRUE)` returns `FALSE`.
+#'
+#' ```r
+#' get_devmode_option <- function(option, default = NULL, on = NULL, message = NULL) {
+#'   if (!in_devmode()) {
+#'     # Dev Mode disabled, act like `getOption()`
+#'     return(getOption(option, default = default))
+#'   }
+#'   # Notify developer
+#'   if (getOption("shiny.devmode.verbose", TRUE) && !is.null(message)) {
+#'     rlang::inform(message = message, .frequency = "regularly", .frequency_id = message)
+#'   }
+#'   # Return Dev Mode default value `on`
+#'   getOption(option, default = on)
+#' }
+#' ```
+#'
+#' The remaining functions in this file are used for author convenience and are not recommended for all situations.
+#' @export
+#' @examples
+#' in_devmode() # TRUE/FALSE?
+#'
+in_devmode <- function() {
+  isTRUE(getOption("shiny.devmode", FALSE)) &&
+    # !testthat::is_testing()
+    !identical(Sys.getenv("TESTTHAT"), "true")
+}
+
+#' @describeIn devmode Temporarily set Shiny Developer Mode and Developer message verbosity
+#' @param code Code to execute with the temporary Dev Mode options set
+#' @export
+#' @examples
+#' # Execute code in a temporary shiny dev mode
+#' with_devmode(TRUE, in_devmode()) # TRUE
+#'
+with_devmode <- function(
+  devmode,
+  code,
+  verbose = getOption("shiny.devmode.verbose", TRUE)
+) {
+  withr::with_options(
+    list(
+      shiny.devmode = devmode,
+      shiny.devmode.verbose = verbose
+    ),
+    code
+  )
+}
+
+
+#' @describeIn devmode If Shiny Developer Mode and verbosity are enabled, displays a message once every 8 hrs (by default)
```

Comment:
I actually think that the default of once per 8 hours is counterproductive for the use cases we have in mind. If people use dev mode, they want to be told about issues in their code.

Another reason that it's problematic is that it makes it difficult to reproduce the problematic message, which makes it hard track down the cause. For example, I was running into some issue for which `rlang::warn()` was emitting a warning once every 8 hours. When I was trying to pinpoint the code that caused it, I had to restart R every time I ran code that triggered the message, because it wouldn't be triggered again without a restart.

Imagine that someone uses (for example) `as.list.reactivevalues` in their code. Then when they run their code, they see the message once, but it doesn't appear when they re-run their code. Then they restart and manage to track it down and fix it. If there are multiple calls to that function, they will only be alerted to the problem once per restart, which makes it more difficult to fix.

Reply:
Deprecation messages are always displayed. 

The only messages that are cached for 8 hours are the messages from `get_devmode_option(key, default)`

Reply:
Ah, makes sense!